File: server/src/scheduler.js
import { runSync } from './connectors/index.js';
import { mapLimit } from './lib/util.js';

/**
 * 定时同步调度：每分钟巡检，对到点的启用数据源触发同步。
 * 不同源之间并发受控（mapLimit），单源失败不影响其他源。
 */
export function startScheduler(store, { syncOnBoot = false, syncIntervalMinutes = 120 } = {}) {
  if (syncIntervalMinutes === 0 && !syncOnBoot) {
    console.log('[scheduler] 定时同步已关闭（KB_SYNC_INTERVAL_MINUTES=0）');
    return { stop: () => {} };
  }

  const tick = async () => {
    const sources = store.listSources().filter((s) => s.enabled);
    const now = Date.now();
    const due = sources.filter((s) => {
      const interval = s.syncIntervalMinutes || syncIntervalMinutes;
      if (!interval || interval <= 0) return false;
      if (!s.lastSync?.finishedAt) return true;
      const elapsed = now - Date.parse(s.lastSync.finishedAt);
      return elapsed >= interval * 60 * 1000;
    });
    if (!due.length) return;
    console.log(`[scheduler] 触发 ${due.length} 个数据源同步`);
    await mapLimit(due, 3, async (s) => {
      try {
        const rec = store.getSource(s.id);
        await runSync(store, rec);
      } catch (e) {
        store.recordSync(s.id, {
          status: 'failed',
          startedAt: new Date().toISOString(),
          finishedAt: new Date().toISOString(),
          durationMs: 0,
          count: 0,
          error: e.message
        });
        console.warn(`[scheduler] 同步失败 ${s.id}: ${e.message}`);
      }
    });
  };

  if (syncOnBoot) {
    setTimeout(tick, 1500);
  }
  const timer = setInterval(tick, 60 * 1000);
  if (timer.unref) timer.unref();

  return {
    stop: () => clearInterval(timer),
    tick
  };
}

File: server_py/kb/scheduler.py
"""定时同步调度：每分钟巡检，对到点的启用数据源触发同步。"""

from __future__ import annotations

import threading
import time
from datetime import datetime

from .connectors import run_sync
from .core import now_iso


class Scheduler:
    def __init__(self, store, sync_interval_minutes: int = 120, sync_on_boot: bool = False):
        self.store = store
        self.interval = int(sync_interval_minutes or 0)
        self.sync_on_boot = sync_on_boot
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    def _due(self, source) -> bool:
        interval = int(source.get("syncIntervalMinutes") or self.interval or 0)
        if interval <= 0:
            return False
        last = (source.get("lastSync") or {}).get("finishedAt")
        if not last:
            return True
        try:
            elapsed = (datetime.now() - datetime.fromisoformat(last.replace("Z", "+00:00")).replace(tzinfo=None)).total_seconds()
        except Exception:  # noqa: BLE001
            return True
        return elapsed >= interval * 60

    def tick(self) -> None:
        due = [s for s in self.store.sources.values() if s.get("enabled") and self._due(s)]
        if not due:
            return
        print(f"[scheduler] 触发 {len(due)} 个数据源同步")
        for s in due:
            try:
                run_sync(self.store, s)
            except Exception as e:  # noqa: BLE001
                self.store.record_sync(s["id"], {"status": "failed", "startedAt": now_iso(),
                                                 "finishedAt": now_iso(), "durationMs": 0, "count": 0, "error": str(e)})
                print(f"[scheduler] 同步失败 {s['id']}: {e}")

    def _loop(self) -> None:
        if self.sync_on_boot:
            time.sleep(1.5)
            self.tick()
        while not self._stop.wait(60):
            self.tick()

    def start(self) -> "Scheduler":
        if self.interval == 0 and not self.sync_on_boot:
            return self
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()
        return self

    def stop(self) -> None:
        self._stop.set()
